File: python3/704.binary-search.py
#
# @lc app=leetcode id=704 lang=python3
#
# [704] Binary Search
#

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left < right:
            mid = left + (right - left) // 2
            # you can not use `target < nums[mid]`
            if target > nums[mid]: left = mid + 1 
            else: right = mid
        return left if nums[left] == target else -1

File: cpp/704.binary-search.cpp
/*
 * @lc app=leetcode id=704 lang=cpp
 *
 * [704] Binary Search
 */

// @lc code=start
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size(), mid;
        while (l < r) {
            mid = l + ((r - l) >> 1);
            if (nums[mid] < target) l = mid + 1;
            else r = mid;
        }

        return (l == nums.size() || nums[l] != target) ? -1 : l;
    }
};
// @lc code=end

